tests/clar_libgit2.h

```diff
@@ -41,6 +41,47 @@
 	} \
 	} while(0)
 
+/**
+ * Thread safe assertions; you cannot use `cl_git_report_failure` from a
+ * child thread since it will try to `longjmp` to abort and "the effect of
+ * a call to longjmp() where initialization of the jmp_buf structure was
+ * not performed in the calling thread is undefined."
+ *
+ * Instead, callers can provide a clar thread error context to a thread,
+ * which will populate and return it on failure.  Callers can check the
+ * status with `cl_git_thread_check`.
+ */
+typedef struct {
+	int error;
+	const char *file;
+	int line;
+	const char *expr;
+	char error_msg[4096];
+} cl_git_thread_err;
+
+#define cl_git_thread_pass(threaderr, expr) cl_git_thread_pass_(threaderr, (expr), __FILE__, __LINE__)
+
+#define cl_git_thread_pass_(__threaderr, __expr, __file, __line) do { \
+	giterr_clear(); \
+	if ((((cl_git_thread_err *)__threaderr)->error = (__expr)) != 0) { \
+		const git_error *_last = giterr_last(); \
```

Comment:
This is the same mechanism that `cl_git_pass` uses.  It's aimed to work as `cl_git_pass(git_function...)`, so it clears any prior errors before executing `git_function`.

Getting rid of this would be annoying when we fail for some reason that _doesn't_ set the git error.  For example, when a function returns `GIT_ITEROVER` or something unexpectedly.  It would be weird to report the last error message which probably has nothing to do with the failure.

It breaks down in this case, where we're `cl_git_pass`ing after executing the thing that failed - sort of like we do here.